pot/pot.go

```diff
@@ -467,75 +489,97 @@ func (t *Pot) each(f func(Val) bool) bool {
 	}
 	_, beg := t.getPos(po)
 	for i := beg; i < len(t.bins); i++ {
-		if !t.bins[i].each(f) {
+		if !t.bins[i].each(consumer) {
 			return false
 		}
 	}
-	return f(t.pin)
+	return consumer(t.pin)
 }
 
-// EachBin iterates over bins of the pivot node and offers iterators to the caller on each
-// subtree passing the proximity order and the size
-// the iteration continues until the function's return value is false
-// or there are no more subtries
-func (t *Pot) EachBin(val Val, pof Pof, po int, f func(int, int, func(func(val Val) bool) bool) bool) {
-	t.eachBin(val, pof, po, f)
+// EachBin iterates over bins relative to the pivot Val node and offers iterators to the caller on each
+// subtree passing the proximity order and the size the iteration continues until the function's return value is false
+// or there are no more subtrees.
+// The order the bins are consumed depends on the bins po with respect to the pivot Val.
+// minProximityOrder gives the caller the possibility of filtering the bins by proximityOrder >= minProximityOrder
+// If pivotVal is the root val it iterates the bin as stored in this pot.
+func (t *Pot) EachBin(pivotVal Val, pof Pof, minProximityOrder int, binConsumer BinConsumer) {
+	t.eachBin(pivotVal, pof, minProximityOrder, binConsumer)
 }
 
-func (t *Pot) eachBin(val Val, pof Pof, po int, f func(int, int, func(func(val Val) bool) bool) bool) {
+func (t *Pot) eachBin(pivotVal Val, pof Pof, minProximityOrder int, consumeBin BinConsumer) {
 	if t == nil || t.size == 0 {
 		return
 	}
-	spr, _ := pof(t.pin, val, t.po)
-	_, lim := t.getPos(spr)
+	valProximityOrder, _ := pof(t.pin, pivotVal, t.po)
+	_, pivotBinIndex := t.getPos(valProximityOrder)
 	var size int
-	var n *Pot
-	for i := 0; i < lim; i++ {
-		n = t.bins[i]
-		size += n.size
-		if n.po < po {
+	var pot *Pot
+	// Consume all bins before the pivotVal bin (or all bins if the pivotVal is the t.pin)
+	// Always filtering bins with proximityOrder < minProximityOrder
+	for i := 0; i < pivotBinIndex; i++ {
+		pot = t.bins[i]
+		size += pot.size
+		if pot.po < minProximityOrder {
 			continue
 		}
-		if !f(n.po, n.size, n.each) {
+		bin := &Bin{
+			ProximityOrder: pot.po,
+			Size:           pot.size,
+			ValIterator:    pot.each,
+		}
+		if !consumeBin(bin) {
 			return
 		}
 	}
-	if lim == len(t.bins) {
-		if spr >= po {
-			f(spr, 1, func(g func(Val) bool) bool {
-				return g(t.pin)
-			})
+	// If pivotBinIndex == len(t.bins), the pivotVal is the t.pin. We consume a virtual bin with max valProximityOrder
+	// and only one element.
+	if pivotBinIndex == len(t.bins) {
+		if valProximityOrder >= minProximityOrder {
+			bin := &Bin{
+				ProximityOrder: valProximityOrder,
+				Size:           1,
+				// Only iterate the pin
+				ValIterator: func(consume ValConsumer) bool {
+					return consume(t.pin)
+				},
+			}
+			consumeBin(bin)
 		}
 		return
 	}
 
-	n = t.bins[lim]
+	pot = t.bins[pivotBinIndex]
 
-	spo := spr
-	if n.po == spr {
+	spo := valProximityOrder
```

Comment:
Since you're at it, please consider giving `spo` an equally descriptive name as the other variables.

Reply:
No idea what this exactly means. It seems an indicator of a proximityOrder where to start finding bins which po higher thatn the pivotBin. Any suggestions?

Reply:
And maybe ther is something I don't understand, but I think spo will be alwais `valProximityOrder + 1`

Reply:
@zelig you mind clearing this up?

Reply:
Anything on this?